_posts/2021-09-06-terraform-uncommon-options.markdown

```diff
@@ -0,0 +1,336 @@
+
+In this blog post, I want to talk about the not-so-very common commands, arguments, and use cases.
+
+
+## [Terraform Console]
+
+This command is my current personal fav, the **console** command. As per documentation it is most commonly used for experimenting with expressions and functions. As an added extra, if the command is invoked in a folder with an active Terraform state it will read it from the configured backend. Inspecting and experimenting with the state with Terraform's console can be fun!
+
+Getting to it is as simple as typing in the command line:
+
+```sh
+$ terraform console
+```
+
+Now we can start "playing" with our state, provided there is something in it that we can fiddle with.
+
+For the sake of this blog post let's create a `main.tf` file with the following contents. The file is also available in my [examples repo].
+
+```hcl
+locals {
+  map = {
+    a = 1
+    b = 2
+    c = 3
+  }
+}
+
+resource "local_file" "private_key" {
+  content  = jsonencode(local.map)
+  filename = "./map.json"
+}
+```
+Needless to say, executing `terraform init` is necessary to pull the plugins and initialize the configured backend if there is any.
+
+Afterwads running `terraform apply` against this template, a JSON encoded file with the content of the local variable `map` will be created in the same folder. That is easy enough.
+
+But what if I needed a YAML encoded file? I am also not sure how it will look alike. I prefer JSON, but I also like to have some comments.
+
+I can check out on the fly by firing up Terraform console in the same folder and just YAML encoding the local `map` variable:
+
+```sh
+$ terraform console
+> yamlencode(local.map)
+<<EOT
+"a": 1
+"b": 2
+"c": 3
+
+EOT
+```
+Voila! We see how our file will look like if we've used the `yamlencode` function instead.
+
+Why stop there, maybe I just need the value of the `b` key?
+
+```sh
+> yamlencode(local.map.b)
+<<EOT
+2
+...
+
+EOT
+```
+
+Swapping the keys and the values is also something that I might want?
+
+```sh
+> yamlencode({for k,v in local.map : v => k })
+<<EOT
+"1": "a"
+"2": "b"
+"3": "c"
+
+EOT
+```
+Or just "trying" out the `try` function to see how it exactly works.
+
+```sh
+> try(yamlencode(local.map.d), 4)
+4
+```
+
+I digress. The flexibility that the `terraform console` provides when testing, troubleshooting, or simply playing around is immense.
+
+## [Shell Autocomplete]
+
+One of the commands that do not appear when invoking terraform via the CLI. Myself having worked with Terraform for quite a while, I stumbled upon this command fairly recently. IT lets you use shell tab-completion. As of writing this blog post, it works with `bash` or `zsh` shells only.
```

Comment:
Your first sentence isn't grammatically correct. Here's how it can change: "This is one of the commands that do not appear when I invoke Terraform via the CLI." Then the first half of your 2nd sentence needs to change too "Having worked with Terraform for quite a while myself, I ...". I'm also not sure if IT needs to be with both capital I and T, or just It